libraries/js/text/regex.effekt

```diff
@@ -1,33 +1,242 @@
+ * ```effekt
+ * val regexWithIndices = regex("(\\w+)", [Global(), GenerateIndices()])
+ * ```
+ */
+def regex(pattern: String): Regex = regex(pattern, [])
 
-def exec(reg: Regex, str: String): Option[Match] =
-   undefinedToOption(reg.unsafeExec(str)) match {
-    case None() => None()
-    case Some(v) => Some(Match(v.matched, v.index))
+
+/**
+ * Constructs a Regex object from a pattern string and optional flags.
+ *
+ * @param pattern The regular expression pattern as a string.
+ * @param flags A List of RegexFlags containing any combination of the following flags:
+ *   - `Global` Global search (find all matches rather than stopping after the first match).
+*       This flags makes the regex object stateful and `exec` will return the next match on each call.
+ *   - `IgnoreCase` Case-insensitive search
+ *   - `Multiline Multi-line search (^ and $ match start/end of each line)
+ *   - `GenerateIndices Generate indices for substring matches. This flag is ignored when using Node
+ *      version < 16.0.0
+ *
+ * @return A Regex object that can be used with the `exec` function.
+ *
+ * Example usage:
+ * ```effekt
+ * val simpleRegex = regex("\\w+")
+ * val complexRegex = regex("(\\w+)\\s(\\d+)", [Global(), GenerateIndices()]))
+ * ```
+ *
+ * Note: The `GenerateIndices` flag is required for extracting substring indices of groups.
+ * If not present, the indices of capture groups will always be `None`.
+ */
+def regex(pattern: String, flags: List[RegexFlags]): Regex = {
+  def showGuard(flag: RegexFlags): String = {
+    flag match {
+      case GenerateIndices() => if (regexIndicesSupported?()) {flag.show()} else {""}
+      case _ => flag.show()
+    }
   }
+  val flags = flags.map{it => it.showGuard()}.join("")
+  internal::unsafeRegex(pattern, flags)
+}
+
+/**
+ * Executes the regex against the given string.
+ *
+ * @param reg The Regex object to execute.
+ * @param str The string to match against.
+ * @return Option[Match] A Some(Match) if a match is found, None otherwise.
+ *
+ * The returned Match object contains:
+ * - matched: The full matched string.
+ * - start: The starting index of the match.
+ * - end: The ending index of the match if `Global` flag is set, otherwise 0
+ * - groups: A list of Group objects for each capturing group.
+ *
+ * Each Group object contains:
+ * - matched: The matched substring for this group.
+ * - index: An Option[Range] with start and end indices if the 'd' flag is set, None otherwise.
+ *
+ * Note: This function is stateful when the global flag 'g' is set on the Regex object.
+ * Subsequent calls with the same Regex and string will return the next match in the string.
+ *
+ * Example usage:
+ * ```effekt
+ * val rx = regex("(\\w+)\\s(\\d+)", "g")
+ * val text = "Hello 123 World 456"
+ * while (true) {
+ *  exec(rx, text) match = {
+ *    Some(m) =>
+ *      println("matched: " ++ m.matched ++ "range: " ++ m.start.show() ++ ", " ++ m.end.show()))
+ *    None() => break
```

Comment:
This code probably doesn't work, right? Could you add this to the tests (verbatim) to make sure it is correct?